pkg/agent/proxy/proxier.go

```diff
@@ -1217,20 +1285,23 @@ func (p *proxier) deleteServiceByIP(serviceStr string) {
 }
 
 func (p *proxier) Run(stopCh <-chan struct{}) {
-	p.once.Do(func() {
-		p.ofClient.RegisterPacketInHandler(uint8(openflow.PacketInCategorySvcReject), p)
-		go p.serviceConfig.Run(stopCh)
-		if p.endpointSliceEnabled {
-			go p.endpointSliceConfig.Run(stopCh)
-			if p.topologyAwareHintsEnabled || p.serviceTrafficDistributionEnabled {
-				go p.nodeConfig.Run(stopCh)
-			}
-		} else {
-			go p.endpointsConfig.Run(stopCh)
-		}
-		p.stopChan = stopCh
-		p.SyncLoop()
-	})
+	p.stopChan = stopCh
+	p.SyncLoop()
+}
+
+func (p *ProxyServer) Run(ctx context.Context) {
+	p.ofClient.RegisterPacketInHandler(uint8(openflow.PacketInCategorySvcReject), p)
+	go p.serviceConfig.Run(ctx.Done())
+	if p.endpointSliceConfig != nil {
+		go p.endpointSliceConfig.Run(ctx.Done())
+	}
+	if p.endpointsConfig != nil {
+		go p.endpointsConfig.Run(ctx.Done())
+	}
+	if p.nodeConfig != nil {
+		go p.nodeConfig.Run(ctx.Done())
+	}
+	p.proxier.GetProxyProvider().Run(ctx.Done())
 }
 
 func (p *proxier) GetProxyProvider() k8sproxy.Provider {
```

Comment:
I wonder about the value of `GetProxyProvider` now, we have a lot of indirection.
Should we embed `k8sproxy.Provider` in our `Proxier` interface and remove this method?

Reply:
In the latest commit:

```go
// Called in cmd/antrea-agent/agent.go to initialize proxy.
type ProxyServer struct {
    ...
    // This could be a single-stack `proxier` defined below, or `metaProxier` defined in third_party/proxy/meta_proxier.go
    proxier k8sproxy.Provider
}

// Single-stack proxy instance
type proxier struct {
   ....
}

// Proxier defines the interface for components that need to query information maintained by the Proxy.
// This is implemented by the single-stack `proxier` defined above, or `metaProxier` defined in third_party/proxy/meta_proxier.go
type Proxier interface {
	GetServiceFlowKeys(serviceName, namespace string) ([]string, []binding.GroupIDType, bool)
	GetServiceByIP(serviceStr string) (k8sproxy.ServicePortName, bool)
}

```